Add support for callables in input and parameters values

arimbr opened the pull request:
Add support for callables in input and parameter values. Those callables should take the context as an argument, and will be resolved at execution time. This is spcially useful to use as input the output of a previous execution, or dynamically set parameters from context.

It also adds the classmethod `get_output_uri` that can be useful and imported from client code.

arimbr commented:
Good question. Internally we use the `add_outputs_to_inputs` argument in `ValohaiSubmitExecutionOperator`. That may be easier to use, but I was afraid that we will end up adding more arguments to the operator for other use cases like updating parameters or inputs at run time. I opted to not add an argumetn that we would have to maintain.

I thought that overriding the Operator will be the case that can handle three advance use cases:
- Add extra inputs (maybe from previous execution)
- Modify parameters at run time
- Modify input names at run time

But maybe that is too complicated for the user. If going so, I would of course add a Guide in the README or wiki over how to do that, together with the examples. I think I need to reflect over this one more time.

jdiprizio commented:
As you said, I think the current approach is great for all the exception use cases you mentioned but may be we could also have the `add_outputs_to_inputs` and document both use cases? Or does that seem complicated to you?

arimbr commented:
Other possibility would be to code this interface that may be clearer for the user (by allowing a dictionary as the input value):

```
train = ValohaiTrainOperator(
    task_id='train_model',
    project_name='tensorflow-example',
    step='Train model (MNIST)',
    dag=dag,
    inputs={
        'test-set-images': {
            'dag_id': preprocess.dag_id,
            'task_id': preprocess.task_id,
            'name': 'mnist-train-images.gz'
        }
    },
    parameters={
        'dropout': 0.9,
        'learning_rate': 0.001,
        'max_steps': 300,
        'batch_size': 200,
    }
)
```

Or even, maybe more clear:

```
train = ValohaiTrainOperator(
    task_id='train_model',
    project_name='tensorflow-example',
    step='Train model (MNIST)',
    dag=dag,
    inputs={
        'test-set-images': ValohaiTrainOperator.get_output('mnist-train-images.gz', task=preprocess),
    },
    parameters={
        'dropout': 0.9,
        'learning_rate': 0.001,
        'max_steps': 300,
        'batch_size': 200,
    }
)
```

arimbr commented:
@jdiprizio At the end I changed the approach and got pretty excited about a new interface that allows to pass a callable as an input/parameter value. I agree with you that is good to support easily passing outputs as inputs. I think now is clear as the input/parameter can be:
1. a constant (string, integer, float,...) that is known before execution time.
2. a callable that will take the context object as parameter and be resolved at execution time.

I believe Case 2 can handle all the cases we have cheated by creating a subclass of ValohaiSubmitExecutionOperator, which will simplify code a lot.

I added an example dag that uses that.

Syntax would be:

```
train = ValohaiSubmitExecutionOperator(
    task_id='train_model',
    project_name='tensorflow-example',
    step='Train model (MNIST)',
    dag=dag,
    inputs={
        'training-set-images': partial(
            ValohaiSubmitExecutionOperator.get_output_uri,
            task=preprocess,
            name='mnist-train-images.gz'),
        'training-set-labels': partial(
            ValohaiSubmitExecutionOperator.get_output_uri,
            task=preprocess,
            name='mnist-train-labels.gz'),
        'test-set-images': partial(
            ValohaiSubmitExecutionOperator.get_output_uri,
            task=preprocess,
            name='mnist-test-images.gz'),
        'test-set-labels': partial(
            ValohaiSubmitExecutionOperator.get_output_uri,
            task=preprocess,
            name='mnist-test-labels.gz'),
    },
    parameters={
        'dropout': 0.9,
        'learning_rate': 0.001,
        'max_steps': 300,
        'batch_size': 200,
    }
)
```

And for dynamic parameters it is possible to define a custom function like:
```
def get_job(context):
    return context['dag_run'].conf['job']
```
And then in parameters:
```
ValohaiSubmitExecutionOperator(
    ...
    parameters = {
        'job': get_job
    }
    ...
)
```
What do you think?
